# Iris Classification

In [1]:
import tensorflow as tf
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

In [2]:
X = load_iris().data
y = load_iris().target

X_train, X_test, y_train, y_test = train_test_split(X, y)

In [3]:
len(X), len(y)

(150, 150)

In [4]:
X[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

In [5]:
y[:5]

array([0, 0, 0, 0, 0])

In [6]:
X[50:55]

array([[7. , 3.2, 4.7, 1.4],
       [6.4, 3.2, 4.5, 1.5],
       [6.9, 3.1, 4.9, 1.5],
       [5.5, 2.3, 4. , 1.3],
       [6.5, 2.8, 4.6, 1.5]])

In [7]:
y[50:55]

array([1, 1, 1, 1, 1])

In [8]:
X[-5:]

array([[6.7, 3. , 5.2, 2.3],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3],
       [5.9, 3. , 5.1, 1.8]])

In [9]:
y[-5:]

array([2, 2, 2, 2, 2])

In [10]:
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((112, 4), (112,), (38, 4), (38,))

In [11]:
X_train[:5]

array([[6.5, 3. , 5.5, 1.8],
       [6.6, 3. , 4.4, 1.4],
       [5.7, 2.6, 3.5, 1. ],
       [5.8, 2.7, 3.9, 1.2],
       [6.7, 2.5, 5.8, 1.8]])

In [12]:
y_train[:5]

array([2, 1, 1, 1, 2])

In [13]:
X_test[:5]

array([[5. , 2. , 3.5, 1. ],
       [5.5, 4.2, 1.4, 0.2],
       [5. , 3.5, 1.6, 0.6],
       [5.5, 2.4, 3.7, 1. ],
       [5.2, 2.7, 3.9, 1.4]])

In [14]:
y_test[:5]

array([1, 0, 0, 1, 1])

In [15]:
# Split trainning and test data

train = tf.data.Dataset.from_tensor_slices( (X_train, y_train) )
train = train.repeat().shuffle(1000).batch(32)

test = tf.data.Dataset.from_tensor_slices( (X_test, y_test) )
test = test.batch(1)

2021-09-05 20:56:18.571516: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [16]:
# Create the model and design the network

model = tf.keras.Sequential([
    # input layer, 4 inputs
    tf.keras.Input( (4,) ),
    # tow hidden layers, 10 neurons each
    tf.keras.layers.Dense( 10, activation=tf.nn.relu ), # activation function is REctivier Linear Unit
    tf.keras.layers.Dense( 10, activation=tf.nn.relu ),
    # output layer, 3 categories, each represent the probability of that category
    tf.keras.layers.Dense( 3 , activation=tf.nn.softmax ), # softmax will turn result to probability
])


In [17]:
# Design the training

model.compile(
    loss='sparse_categorical_crossentropy', # method to caculate how wrong the answer is, 
                                            # then feed to back propagation to modify the weight
    optimizer='adam', # a sort of gradient descent, how much the weight should be changed
    metrics=['accuracy'] # how the training is progressing
)


In [18]:
# Training and evaluating

model.fit(
    train,
    validation_data=test,
    steps_per_epoch=150, # for each big roop of training data
                         # you should do look at (150 examples * 32 batches) flowers
    epochs=10            # do 10 big loops
)

Epoch 1/10
 47/150 [========>.....................] - ETA: 0s - loss: 1.3671 - accuracy: 0.3457 

2021-09-05 20:59:08.535216: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)


150/150 [==============================] - 1s 2ms/step - loss: 1.0987 - accuracy: 0.5523 - val_loss: 0.8611 - val_accuracy: 0.6579
Epoch 2/10
150/150 [==============================] - 0s 1ms/step - loss: 0.6944 - accuracy: 0.8017 - val_loss: 0.5416 - val_accuracy: 0.8947
Epoch 3/10
150/150 [==============================] - 0s 1ms/step - loss: 0.4057 - accuracy: 0.9504 - val_loss: 0.3214 - val_accuracy: 0.9474
Epoch 4/10
150/150 [==============================] - 0s 1ms/step - loss: 0.2453 - accuracy: 0.9681 - val_loss: 0.2198 - val_accuracy: 0.9474
Epoch 5/10
150/150 [==============================] - 0s 1ms/step - loss: 0.1701 - accuracy: 0.9712 - val_loss: 0.1746 - val_accuracy: 0.9474
Epoch 6/10
150/150 [==============================] - 0s 1ms/step - loss: 0.1251 - accuracy: 0.9698 - val_loss: 0.1659 - val_accuracy: 0.9474
Epoch 7/10
150/150 [==============================] - 0s 1ms/step - loss: 0.0983 - accuracy: 0.9773 - val_loss: 0.1584 - val_accuracy: 0.9474
Epoch 8/10
150/15

The result above:
- `- loss: 1.0987 - accuracy: 0.5523` is from the train dataset
- `- val_loss: 0.8611 - val_accuracy: 0.6579` is from the testset

In [19]:
# Use the model

predict_true_labels = ["setosa", "versicolor", "virginica"]
predict_X = [
    [5.1, 3.3, 1.7, 0.5],
    [5.9, 3.0, 4.2, 1.5],
    [6.9, 3.1, 5.4, 2.1]
]

predictions = model.predict(predict_X)

for pred_dict, expected in zip(predictions, predict_true_labels):
    print('pred_dict:', pred_dict)  # a list with 3 predicted probabilities for 3 categories as we designed
    predicted_index = pred_dict.argmax()
    probability = pred_dict.max()
    predicted = load_iris().target_names[predicted_index]
    
    tick_cross = "🙆‍" if predicted == expected else "🙅‍♂️"
    print(f"{tick_cross} Prediction is '{predicted}' ({100 * probability: .1f}%), expected {expected}")

pred_dict: [9.8634481e-01 1.3655217e-02 1.1081977e-10]
🙆‍ Prediction is 'setosa' ( 98.6%), expected setosa
pred_dict: [0.00266769 0.98289746 0.01443486]
🙆‍ Prediction is 'versicolor' ( 98.3%), expected versicolor
pred_dict: [6.8847903e-06 8.1386946e-02 9.1860622e-01]
🙆‍ Prediction is 'virginica' ( 91.9%), expected virginica


# Image Analysis

## Introduction

An image can bee treated as a large grid binary number (between 0 and 255):

```
# a gray picture, each number is a brightness

105 102 100 ...
103 109 102 ...
99  101 100 ...
... ... ... ...
```

## Kernel Convolution

- You can create a kernel which defines a filter to be applied to the image:

```
# sharpen

kernel = [
    [0, -1, 0],
    [-1, 5, -1],
    [0, -1, 0]
]
```

- Depending on the values in the kernel different filtering operations will be performed. The most common are:
    - sharpen (edge detection)
    - blur
    - edge detection
    
- The values of the kernels are created by mathmetical analysis and are generally fixed

## Example

1. Origin pictur:

```
105 102 100 ...
103 109 102 ...
99  101 100 ...
... ... ... ...
```

2. 3 * 3 sharpen kernal

```
  0  -1   0
 -1   5  -1
  0  -1   0
```

3. After

```
# step 1. apply the kernel
   0 -102    0
-103  545 -102
   0 -101    0
```

```
# step 2. calculate
Nan   Nan   Nan
Nan   137   Nan
Nan   Nan   Nan
```

```
# step 3. move the kernel by 1 grid, get the next value
Nan   Nan   Nan
Nan   137     ?
Nan   Nan   Nan
```

```
# Note: 0 can be added outside the edge to calculate the edge grid
```

# Handwriting Recognition

## Convolution Neural Network

- Create abstract feature detectors automatically;


## MINST Dataset



## Design the CNN

1. Convolutional Layer #1
2. Pooling Layer #1
3. Convolutional Layer #2
4. Pooling Layer #2
5. Dens Layer #1
6. Dens Layer #2 (Logist Layer)